config/config.go

```diff
@@ -31,6 +35,13 @@ type AddrsFactory = bhost.AddrsFactory
 // NATManagerC is a NATManager constructor.
 type NATManagerC func(inet.Network) bhost.NATManager
 
+type BasicRouting interface {
+	routing.ContentRouting
+	routing.PeerRouting
+}
+
+type RoutingC func(host.Host) (BasicRouting, error)
```

Comment:
What if we had this return `routing.PeerRouting` router and then used a type switch for everything else? Basically, peer routing is the minimum functionality but, if we have content routing, we can construct a discovery service based on it. In the future, we could check to see if the router supported *other* interfaces (i.e., it may directly implement the discovery interface).

Reply:
Actually, it's not so bad; we can pass `PeerRouting` as a minimum signifying that we want a routed host.  The additional functionality is semi optional, and as you say maybe we'll implement the discovery interface directly from the DHT one day.

Reply:
Reduced to `PeerRouting` + cast per your suggestion in https://github.com/libp2p/go-libp2p/pull/454/commits/c91a0d34406fbb70c689e1b3e8ddc06270cb0cca